### DISCUSSION TITLE
How to test actual workflow execution

### DISCUSSION THREAD

**Discussion by ankit-joinwal** [camunda-community-hub/zeebe-test-container#194] [Q&A] [answered]
Hello,

We have a CI/CD pipeline to manage workflow definitions (.bpmn). The micro-services which are part of a workflow process have their own CI/CD pipelines.

Question - what are the options for testing the workflows end-to-end? Eg- deploy the workflow to Zeebe, start process instance so it interacts with actual micro-services and ensure the execution succeeds. Doing all this in a CI/CD pipeline.

Another question is what is the use case for zeebe-test-container ? when would we want to use it?

**Comment by npepinpe** [comment, marked as answer]
Hi ankit,

`zeebe-test-container` is a library mostly for integration testing. Ideally, you would unit test the logic of your processes with something like `camunda-bpm-assert`. I understand there's nothing right now like that in the Zeebe ecosystem, but we are working on an upcoming library for this targeting Q2 next year: see https://github.com/camunda-cloud/camunda-cloud-testing/.

In the meantime, there are some community projects which provide lightweight alternatives for unit testing your workflows. See https://github.com/camunda-community-hub/eze.

That said, you can test your processes with `zeebe-test-container`, but as with most integration tests, you won't have fine grained control. If you wanted to test a process, as you mentioned, you would deploy it, then create an instance (or publish a message, anything to trigger the start of an instance). You'd also have to be running your workers somewhere already - this can be in the same JVM, but it can also be other containers as well.

Verification would be ensuring your workers are called with the expected input, that a message triggers the right event, etc. But as mentioned, you won't have very fine grained control.

The use case of the library is then more about integration testing and/or end-to-end testing. While you should always unit test your worker code, and in the future unit test the process definition itself, you will still need integration tests that check the workers are orchestrated together properly. You won't have fine grained control to test all possible failure paths of course, but it should still give you increased confidence by testing the general cases.

The library is also often used by exporter authors, or by those building application based on the exported stream (e.g. Operate, Tasklist, Zeeqs, etc.), again for the integration tests. You don't test every single thing at that level, but you do want some basic tests to make sure it works against a real Zeebe, which is what the library provides.

**Comment by npepinpe** [comment]
A short update, you can now use this project in combination with [zeebe-process-test](https://github.com/camunda-cloud/zeebe-process-test/). See the [README](https://github.com/camunda-community-hub/zeebe-test-container#zeebe-process-test-compatibility) for more on how to do this.

That said, in most cases you should just use `zeebe-process-test` directly. If all you want to do is test your process definitions, then use that. But you might want to use `zeebe-test-container` if:

- You're developing Zeebe itself
- You're developing a Zeebe gateway interceptor and want to write integration tests for it
- You're developing a Zeebe exporter and want to write integration tests for it
- You use advanced features of Zeebe (e.g. the rebalancing API, a fixed partitioning scheme with multiple partitions, etc.) which are only available in the full application, and want to write integration tests against a workflow which involves these features and how they interact with other resources (e.g. BPMN processes/DMN models, workers, etc.)
